core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/pythonworker/PythonProxyClient.scala

```diff
@@ -0,0 +1,142 @@
+package edu.uci.ics.amber.engine.architecture.pythonworker
+
+import edu.uci.ics.amber.engine.architecture.pythonworker.WorkerBatchInternalQueue.{
+  ControlElement,
+  ControlElementV2,
+  DataElement
+}
+import edu.uci.ics.amber.engine.common.ambermessage.InvocationConvertUtils.{
+  controlInvocationToV2,
+  returnInvocationToV2
+}
+import edu.uci.ics.amber.engine.common.ambermessage.{PythonControlMessage, _}
+import edu.uci.ics.amber.engine.common.rpc.AsyncRPCClient.{ControlInvocation, ReturnInvocation}
+import edu.uci.ics.amber.engine.common.virtualidentity.ActorVirtualIdentity
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+import edu.uci.ics.texera.workflow.common.tuple.schema.Schema
+import org.apache.arrow.flight._
+import org.apache.arrow.memory.{BufferAllocator, RootAllocator}
+import org.apache.arrow.vector.VectorSchemaRoot
+
+import scala.collection.mutable
+
+class PythonProxyClient(portNumber: Int)
+    extends Runnable
+    with AutoCloseable
+    with WorkerBatchInternalQueue {
+
+  val allocator: BufferAllocator =
+    new RootAllocator().newChildAllocator("flight-client", 0, Long.MaxValue)
+  val location: Location = Location.forGrpcInsecure("localhost", portNumber)
+
+  private val MAX_TRY_COUNT: Int = 3
+  private val WAIT_TIME_MS = 500
+  private var flightClient: FlightClient = _
+
+  override def run(): Unit = {
+    establishConnection()
+    mainLoop()
+  }
+
+  def establishConnection(): Unit = {
+    var connected = false
+    var tryCount = 0
+    while (!connected && tryCount < MAX_TRY_COUNT) {
+      try {
+        println("trying to connect to " + location)
```

Comment:
maybe better to just create a logger for this class and use the logger?

Reply:
changed to logger, which passed from WorkflowActor.